aws/lambda/pytorch-auto-revert/pytorch_auto_revert/signal.py

```diff
@@ -0,0 +1,291 @@
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import List, Optional, Set, Union
+
+
+class SignalStatus(Enum):
+    """signal status enum"""
+
+    PENDING = "pending"
+    SUCCESS = "success"
+    FAILURE = "failure"
+
+
+# data classes:
+
+
+@dataclass
+class AutorevertPattern:
+    """
+    Represents an autorevert pattern detected in a signal.
+    """
+
+    pattern_detected: bool
+    workflow_name: str
+    newer_commits: List[str]
+    older_commit: str
+    # failed_job_names: List[str]  # TODO: Uncomment when needed
+    # failure_rule: str              # TODO: Uncomment when needed
+    # job_name_base: str             # TODO: Uncomment when needed
+
+
+@dataclass
+class RestartCommits:
+    """
+    Represents an intent to restart a specific set of commits on the signal.
+    """
+
+    commit_shas: Set[str]
+
+
+class SignalEvent:
+    """A single observation contributing to a Signal on a given commit.
+
+    Represents one job/test/classification-derived event with a status and
+    start/end timestamps used to reason about ordering and patterns.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        status: SignalStatus,
+        started_at: datetime,
+        ended_at: Optional[datetime] = None,
+    ):
+        self.name = name
+        self.status = status
+        self.started_at = started_at
+        self.ended_at = ended_at
+
+    @property
+    def is_pending(self) -> bool:
+        return self.status == SignalStatus.PENDING
+
+    @property
+    def is_success(self) -> bool:
+        return self.status == SignalStatus.SUCCESS
+
+    @property
+    def is_failure(self) -> bool:
+        return self.status == SignalStatus.FAILURE
+
+
+class SignalCommit:
+    """All events for a single commit, ordered oldest → newest by start time."""
+
+    def __init__(self, head_sha: str, events: List[SignalEvent]):
+        self.head_sha = head_sha
+        # enforce events ordered by time, oldest first
+        self.events = sorted(events, key=lambda e: e.started_at) if events else []
+
+    def events_by_status(self, status: SignalStatus) -> List[SignalEvent]:
```

Comment:
this property is used in multiple places, and it is creating a new list.

Either:
1 - cache by status; (faster if reused more frequently)
2 - Return a generator instead; (prefered IMO)